sound/soc/sof/ipc4-topology.c

```diff
@@ -2002,12 +2089,20 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 {
 	struct snd_sof_widget *src_widget = sroute->src_widget;
 	struct snd_sof_widget *sink_widget = sroute->sink_widget;
+	struct snd_sof_widget *src_pipe_widget = src_widget->spipe->pipe_widget;
+	struct snd_sof_widget *sink_pipe_widget = sink_widget->spipe->pipe_widget;
 	struct sof_ipc4_fw_module *src_fw_module = src_widget->module_info;
 	struct sof_ipc4_fw_module *sink_fw_module = sink_widget->module_info;
+	struct sof_ipc4_pipeline *src_pipeline = src_pipe_widget->private;
+	struct sof_ipc4_pipeline *sink_pipeline = sink_pipe_widget->private;
 	struct sof_ipc4_msg msg = {{ 0 }};
 	u32 header, extension;
 	int ret;
 
+	/* no route set up if chain DMA is used */
+	if (src_pipeline->use_chain_dma || sink_pipeline->use_chain_dma)
+		return 0;
```

Comment:
question: in reality we will have two pipelines and both have the use_chain_dma set, right? And afaik it is an error if any side does not have this flag, we would not be here. I think.
If we have a path which should use ChainDMA then _all_ pipelines in it must have the flag, so it is enough to check in the src, the test on the sink will never happen, right?

Reply:
Yes its an error if only one of the pipelines has use_chain_dma attribute, but I think its better to avoid misconfiguration at this phase and produce a clear error at trigger phase (see my reply to your earlier comment here: https://github.com/thesofproject/linux/pull/3953#discussion_r1033267616

Reply:
Right, what I meant is that if we are here then we have already went through the sanity check of the pipeline path that all (or none of them) have the flag set, thus if any of the pipelines in the path have the flag then for sure all will have it, no?

Reply:
Have we? The order the each callback is called is not completely clear to me, but I would have expected route_setup to be before trigger. However, if its not then sure I'll remove one of the checks.

Reply:
hrm, so you do double sanity check?
You are right that the route setup should be prior to trigger, so in trigger you should already have a route which either uses ChainDMA or  not use ChainDMA as you already failed the erroneous mixed use?

Reply:
The check is in trigger, because there I will anyway have to pull everything together process both pipelines in the relevant pipeline_list, but yes the check could be here too.  Should I move it here, and simplify the trigger function a little? Or is it still better that as much as possible of the chain dma code is in one place in trigger function?

Reply:
Changed it after all to have the check in sof_ipc4_route_setup() and to simplify the sof_ipc4_trigger_pipelines() a bit.